src/Tribe/Assets.php

```diff
@@ -0,0 +1,367 @@
+<?php
+// Don't load directly
+if ( ! defined( 'ABSPATH' ) ) {
+	die( '-1' );
+}
+
+/**
+ * Class used to register and enqueue assets across our plugins
+ */
+class Tribe__Assets {
+	/**
+	 * Static Singleton Holder
+	 *
+	 * @var self|null
+	 */
+	protected static $instance;
+
+	/**
+	 * Stores all the Assets and it's configurations
+	 *
+	 * @var array
+	 */
+	private $assets = array();
+
+	/**
+	 * Static Singleton Factory Method
+	 *
+	 * @return self
+	 */
+	public static function instance() {
+		if ( empty( self::$instance ) ) {
+			self::$instance = new self;
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Register the Methods in the correct places
+	 */
+	private function __construct() {
+		// Hook the actual registering of
+		add_action( 'init', array( $this, 'action_maybe_register' ), 1 );
+	}
+
+	/**
+	 * Register the Assets on the correct hooks
+	 *
+	 * @return void
+	 */
+	public function action_maybe_register() {
+		foreach ( $this->assets as $asset ) {
+			if ( 'js' === $asset->type ) {
+				wp_register_script( $asset->slug, $asset->url, $asset->deps, $asset->version, $asset->in_footer );
+			} else {
+				wp_register_style( $asset->slug, $asset->url, $asset->deps, $asset->version, $asset->media );
+			}
+
+			// If we don't have an action we don't even register the action to enqueue
+			if ( ! is_string( $asset->action ) ) {
+				continue;
+			}
+
+			// Now add an action to enqueue the registered assets
+			add_action( $asset->action, array( $this, 'enqueue' ), $asset->priority );
+		}
+	}
+
+	public function enqueue() {
+		foreach ( $this->assets as $asset ) {
+			// Skip if we are not on the correct filter
+			if ( current_filter() !== $asset->action ) {
+				continue;
+			}
+
+			// If here is no action we are just registering this asset
+			if ( ! is_string( $asset->action ) ) {
+				continue;
+			}
+
+			/**
+			 * Allows developers to hook-in and prevent an asset from been loaded
+			 *
+			 * @param bool   $enqueue If we should enqueue or not a given asset
+			 * @param object $asset   Which asset we are dealing with
+			 */
+			$enqueue = apply_filters( 'tribe_asset_enqueue', true, $asset );
+
+			/**
+			 * Allows developers to hook-in and prevent an asset from been loaded
+			 *
+			 * Note: When you pass callables on the `$asset->filter` argument this will be hooked here
+			 *
+			 * @param bool   $enqueue If we should enqueue or not a given asset
+			 * @param object $asset   Which asset we are dealing with
+			 */
+			$enqueue = apply_filters( 'tribe_asset_enqueue_' . $asset->slug, $enqueue, $asset );
+
+			if ( ! $enqueue ) {
+				continue;
+			}
+
+			if ( 'js' === $asset->type ) {
+				wp_enqueue_script( $asset->slug );
```

Comment:
Since we're only passing the asset slug to `wp_enqueue_script` here this will fail for any assets registered later than `init` (ie, when `action_maybe_register` runs), but that limits our ability to be lazy and register stuff on demand - anything set up with a `tribe_asset()` call later than this point won't make it to the browser.

What do you think about re-running `wp_register_script()` here? We could possibly track which have already been registered and only do this as needed.

Ditto for registration of styles, below.